src/reactive/prometheus-apt-exporter.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/python3
+"""Installs and configures prometheus-apt-exporter."""
+import hashlib
+import json
+import os
+import shutil
+import subprocess
+from pathlib import Path
+from zipfile import BadZipFile, ZipFile
+
+from charmhelpers.contrib.charmsupport import nrpe
+from charmhelpers.core import hookenv, host
+
+from charms.layer import snap
+from charms.reactive import (
+    endpoint_from_flag,
+    hook,
+    remove_state,
+    set_state,
+    when,
+    when_all,
+    when_any,
+    when_not,
+    when_not_all,
+)
+
+
+DASHBOARD_PATH = os.getcwd() + "/files/grafana-dashboards"
+SNAP_NAME = "prometheus-apt-exporter"
+#SVC_NAME = "snap.prometheus-apt-exporter.daemon"
+SVC_NAME = "snap.prometheus-apt-exporter.apt-exporter.service"
+PORT_NUMBER = "8089"
+
+
+@when("juju-info.connected")
+@when_not_all("apt-exporter.installed", "apt-exporter.started")
+def install_packages():
+    """Installs the prometheus-apt-exporter snap."""
+    hookenv.status_set("maintenance", "Installing software")
+    config = hookenv.config()
+    channel = config.get("snap_channel")
+    snap.install(SNAP_NAME, channel=channel, force_dangerous=False)
+    subprocess.check_call(["snap", "connect", "prometheus-apt-exporter:apt-exporter-files"])
+    hookenv.status_set("active", "Exporter installed and connected")
+    hookenv.open_port(PORT_NUMBER)
+    set_state("apt-exporter.installed")
+
+
+@hook("upgrade-charm")
+def upgrade():
+    """Reset the install state on upgrade, to ensure resource extraction."""
+    hookenv.status_set("maintenance", "Charm upgrade in progress")
+    remove_state("apt-exporter.installed")
+    remove_state("apt-exporter.started")
+    remove_state("apt-exporter.dashboard-registered")
+    update_dashboards_from_resource()
+    register_grafana_dashboards()
+
+
+@when_not("apt-exporter.started")
+@when_any("apt-exporter.installed", "config.changed")
+def start_snap():
+    """Configure snap.prometheus-apt-exporter.apt-exporter service."""
+    if not host.service_running(SVC_NAME):
+        hookenv.status_set("maintenance", "Service is down, starting")
+        hookenv.log("Service {} is down, starting...".format(SVC_NAME))
+        host.service_start(SVC_NAME)
+        hookenv.status_set("active", "Service started")
+        hookenv.log("start_snap() Service started")
+    else:
+        hookenv.status_set("active", "Ready")
+        set_state("apt-exporter.started")
+        hookenv.log("start_snap() apt-exporter.started")
+
+    update_dashboards_from_resource()
+
+    hookenv.log("Installed and set flag apt-exporter.started")
+
+
+@when("config.changed.snap_channel")
+def snap_channel_changed():
+    """Remove the state apt.exporter.installed if the snap channel changes."""
+    remove_state("apt-exporter.installed")
+    remove_state("apt-exporter.started")
```

Comment:
same issue as above (and below) clearing the flags does not ensure immediate execution of decorated functions.